### ISSUE TITLE
Traits

### ISSUE THREAD

**Issue by lawremi** [RConsortium/OOP-WG#34]
Following from Hadley's comment in #31,

Traits sound like a Java "interface", i.e., a method contract, except somewhat complicated by multiple dispatch. For example, for a two argument signature, the contract could specify:
```R
MyTrait <- newTrait(signature("MyTrait", "ANY"), 
                    signature("SomeOtherClassOrTrait", "MyTrait"))
```

In practice, it's been sufficient to rely on duck typing, although it would be nice to have automatic checks. A nice thing about duck typing is that packages can define methods without even knowing about a trait, and interoperate by informal convention. Another potential feature would be automatically identifying whether a class conforms to a trait, without actually declaring it. Runtime checks could use that to fail early.

**Comment by hadley** [comment]
Maybe all I want is to be able to dynamically tell if an object has certain methods defined (possibly excluding default methods?). Then a trait is just a run-time check that an object implements all needed methods — the primary advantage being that you can deliver a clear error message at the start of a function, rather than erroring part way through because a required method is absent.

**Comment by lawremi** [comment]
Sounds like you're describing exactly what I described in the last couple sentences, and I agree that would be the most useful feature.

**Comment by hadley** [comment]
Ooops, yeah. Although I think my point is that the OO system doesn't need to define traits, as long as it provides an easy and efficient way to tell which methods are registered.

**Comment by lawremi** [comment]
That's right. Traits would just make those checks more convenient, as the set of methods would defined in one place, and potentially checked in multiple places.

**Comment by lawremi** [comment]
That's right.

**Comment by hadley** [comment]
So something like this in essence:

```R
library(R7)

# Object can be used with for-loop
for_trait <- new_trait(c("length", "[["))

foo <- new_class("foo", constructor = function(...) {
  check_trait(for_trait, new_object(...))
})
```

**Comment by yutannihilation** [comment]
One of the considerations to introducing a trait feature is how to deal with the name collision. I heard Rust's trait is a bit different from other languages (e.g., PHP and Scala, though I know nothing about the details...) in that Rust allows it and requires an explicit syntax when calling any of the methods.

https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#fully-qualified-syntax-for-disambiguation-calling-methods-with-the-same-name

For an example of another approach, PHP raises error in the case. It requires explicitly choose which one to use, and it also provides `as` syntax for renaming the method name.

> If two Traits insert a method with the same name, a fatal error is produced, if the conflict is not explicitly resolved. 

https://www.php.net/manual/en/language.oop5.traits.php#language.oop5.traits.conflict

**Comment by hadley** [comment]
@yutannihilation I don't _think_ we have to worry about that problem because R7 uses generic function OO (i.e. methods belong to generics) not message passing OO (i.e. methods belong to classes).

**Comment by yutannihilation** [comment]
Ah, thanks, makes sense. Considering the name collision is what can happen on defining a generic function, we don't need to worry about it here.

**Comment by hadley** [comment]
Julia equivalent: https://docs.julialang.org/en/v1/manual/interfaces/

**Comment by hadley** [comment]
Interesting idea from Julia: https://www.ahsmart.com/pub/holy-traits-design-patterns-and-best-practice-book/#example_1_base_iteratorsize — traits are not baked in to the OO system, but are a pattern implemented on top of it.

**Comment by lawremi** [comment]
I am confused by the code above, is it supposed to be:

```
method(length, foo) <- function(x) length(x@data)
```
Instead of:
```
method(`[[`, length) <- function(x, i) length(x@data)
```
?

Another way to think about a trait is as a dynamic union of all classes for which methods exist for a set of generics (really for specific arguments of specific generics). That would make it possible for someone writing a method to define a custom trait that meets the specific requirements of that method definition, and the system would automatically find all implementing classes and define methods for them like it would for a union. The complication is that it would need to be dynamic --- if a new or existing class comes to meet the requirements of the trait, the system would need to detect that and define the associated methods. Luckily there would be no performance hit at runtime; it would just require some extra logic when registering methods.

**Comment by sounkou-bioinfo** [comment]
> Maybe all I want is to be able to dynamically tell if an object has certain methods defined (possibly excluding default methods?). Then a trait is just a run-time check that an object implements all needed methods — the primary advantage being that you can deliver a clear error message at the start of a function, rather than erroring part way through because a required method is absent.

Mostly derivative of this issue (and this specific point) among other things,  interfaces following Golang mental model and "traits" implemented with S7 here https://github.com/sounkou-bioinfo/s7contract
